api/src/glfs-fops.c

```diff
@@ -6319,10 +6326,11 @@ GFAPI_SYMVER_PUBLIC_DEFAULT(glfs_xreaddirplus_get_stat, 3.11.0)
 struct stat *
 pub_glfs_xreaddirplus_get_stat(struct glfs_xreaddirp_stat *xstat)
 {
+    xlator_t *this = THIS;
```

Comment:
Please, don't blindly apply the comments and wait until someone else checks if everything is right. It's the third time I raise this issue in this patch: **`THIS` cannot be used in public functions before calling `__GLFS_ENTRY_VALIDATE_FD()`**.

In this particular case the function signature doesn't include the required fields to correctly initialize `THIS`, so the only good solution here is to replace `this->name` in line 6333 by a string constant.

I haven't checked if there are more cases like this one.

Reply:
Hi @xhernandez - just out of curiosity:
```
    DECLARE_OLD_THIS;
    __GLFS_ENTRY_VALIDATE_FD(glfd, invalid_fs);
```

This is seen tens of times. And DECLARE_OLD_THIS is actually calling 'THIS' and assigning it to old_THIS.
- Is that wrong?
- It's clearly redundant to the same assignment done @ __GLFS_ENTRY_VALIDATE_FD macro.

Reply:
> Hi @xhernandez - just out of curiosity:
> 
> ```
>     DECLARE_OLD_THIS;
>     __GLFS_ENTRY_VALIDATE_FD(glfd, invalid_fs);
> ```
> 
> This is seen tens of times. And DECLARE_OLD_THIS is actually calling 'THIS' and assigning it to old_THIS.
> 
> * Is that wrong?

It's not wrong if `THIS` contains a meaningful value or `old_THIS` is not used anywhere else beyond `__GLFS_EXIT_FS`. However in most (all ?) of the cases this is done in public API functions, which means that they can be called from any thread, and given that `THIS` accesses a thread-local variable, it will most probably be uninitialized, so at the very minimum, those calls are unnecessary IMO.

> * It's clearly redundant to the same assignment done @ __GLFS_ENTRY_VALIDATE_FD macro.

The main issue here is that `old_THIS` is declared by `DECLARE_OLD_THIS` macro, so it's needed to avoid compilation errors when `__GLFS_ENTRY_VALIDATE_FD` is used (it assigns to `old_THIS` but doesn't declare it). Probably `DECLARE_OLD_THIS` should have been limited to declare the variable, without initializing it, but most of the code already depends on this right now, so we can't change it easily.

In this particular case, `DECLARE_OLD_THIS` is saving a potential garbage value of `THIS`, but since it's not used, even if unnecessary, it's not a big problem. The big problem is to take a potentially uninitialized `THIS` an use it inside the function, as in this case. That's why I wrote the comment.

Reply:
In the case of priv_glfs_process_upcall_event, it's a private function, but still looks different than the others, who, from what I can tell, do call __GLFS_ENTRY_VALIDATE_FS / __GLFS_ENTRY_VALIDATE_FD (which assign THIS):
```

    DECLARE_OLD_THIS;

    gf_msg_debug(THIS->name, 0, "Upcall gfapi callback is called");

    __GLFS_ENTRY_VALIDATE_FS(fs, err);

```

This is the only one I've seen that is different, and thus we can probably move DECLARE_OLD_THIS to actually just declare.

Reply:
I'm not sure if the `gf_msg_debug()` shouldn't be moved after `__GLFS_ENTRY_VALIDATE_FS()` in this case.

Since `DECLARE_OLD_THIS` is defined in glusterfs/globals.h, I assumed it's used globally in the code, however it seems its usage is limited to gfapi and one instance in rpc-clnt.c, so maybe it's easier than I thought to modify it.

Reply:
The rpc-clnt one should go away via https://github.com/gluster/glusterfs/pull/3146/commits/6a4df415cafa8fa055cd5211d8d19452bc10fd80